regression_detector.py

```diff
@@ -23,13 +24,62 @@ def get_default_output_path(bm_name: str) -> str:
     return os.path.join(output_path, fname)
 
+def generate_regression_yaml(control, treatment) -> None:
+    # Write result to $REPO_DIR/.userbenchmark/<userbenchmark-name>/regression-<time>.json
+    assert control["name"] == treatment["name"], f'Expected the same userbenchmark name from metrics files, \
+                                                but getting {control["name"]} and {treatment["name"]}.'
+    bm_name = control["name"]
+    detector = importlib.import_module(f"userbenchmark.{bm_name}.regression_detector").run
+
+    # Process control and treatment to include only shared keys
+    filtered_control_metrics = {}
+    control_only_metrics = {}
+    filtered_treatment_metrics = {}
+    treatment_only_metrics = {}
+    for control_name, control_metric in control["metrics"].items():
+        if control_name in treatment["metrics"]:
+            filtered_control_metrics[control_name] = control_metric
+        else:
+            control_only_metrics[control_name] = control_metric
+    for treatment_name, treatment_metric in treatment["metrics"].items():
+        if treatment_name in control["metrics"]:
+            filtered_treatment_metrics[treatment_name] = treatment_metric
+        else:
+            treatment_only_metrics[treatment_name] = treatment_metric
+    control["metrics"] = filtered_control_metrics
+    treatment["metrics"] = filtered_treatment_metrics
+    assert filtered_control_metrics.keys() == filtered_treatment_metrics.keys()
+
+    # Local file comparison, return the regression detection object
+    result = call_userbenchmark_detector(detector, control, treatment)
+    if result or control_only_metrics or treatment_only_metrics:
+        if not args.output:
+            args.output = get_default_output_path(bm_name)
+        # dump result to yaml file
+        result_dict = asdict(result)
+        result_dict["control_only_metrics"] = control_only_metrics
+        result_dict["treatment_only_metrics"] = treatment_only_metrics
+        # create the output directory if doesn't exist
+        output_dir = Path(os.path.dirname(args.output))
+        output_dir.mkdir(parents=True, exist_ok=True)
+        output_yaml_str = yaml.safe_dump(result_dict, sort_keys=False)
+        with open(args.output, "w") as ofptr:
+            ofptr.write(output_yaml_str)
+        print(output_yaml_str)
+        print(f"Wrote above yaml to {args.output}.")
+    else:
+        print(f"No performance signal detected between file {args.control} and {args.treatment}.")
+
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
     # Local metrics file comparison
     parser.add_argument("--control", default=None, help="The control group metrics file for comparison.")
     parser.add_argument("--treatment", default=None, help="The treatment metrics file for comparison.")
 
     # S3 metrics file comparison
+    parser.add_argument("--file", "-f", help="Name of the control metrics JSON file. When provided with "
```

Comment:
If we are comparing with the previous JSON file, the metrics file here should be treatment, not control group, correct?

I think a better interface is to directly use `--treatment` instead, and if `--control` is None, by default we should search for the most recent JSON file as the control group.

If `--treatment` specifies a local file path, we read from local file, otherwise, we try to read from Amazon S3.

Reply:
oh yes, i meant treatment here and not control, the code has it as the treatment.

I can rework this to rely on treatment for clarity